[BREAKING] Throw error if function parameters (return) use default data location.

chase1745 opened the pull request:
Fixes #3402 

Only warns if the contract is a library, but can be changed to all contracts if needed.
Throws Type Error for 0.5.0.

chase1745 committed:
Warn (error - 0.5) if function parameter (return as well) uses default data location.

chriseth commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -367,8 +367,29 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
-					if (_variable.isCallableParameter())
-						typeLoc = DataLocation::Memory;
+					if (_variable.isCallableParameter()) {
+                        typeLoc = DataLocation::Memory;
```
Please indent using tabs only.
Please put `{` on a line of its own.

Coding style: https://github.com/ethereum/solidity/blob/develop/CODING_STYLE.md

chase1745 replied:
My apologies on this! Should be fixed now.

chriseth commented on test/libsolidity/syntaxTests/functionTypes/warn_function_return_parameters_no_data_location.sol:
```diff
@@ -0,0 +1,5 @@
+    function f() private pure returns(uint[]) {}
+}
+// ----
+// Warning: (50-56): Parameter is declared as memory. Use an explicit "memory" keyword to silence this warning.
```
Please add a test that shows that this warning can actually be removed by adding `memory`.

Please also add testst for external and internal functions and functions using storage parameters, both in contracts and in libraries.

chase1745 replied:
I have added these tests for all types of function visibility as well as for contracts and libraries. I also added tests showing the warning is removed by adding a storage location.

chriseth commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -367,8 +367,29 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
+                        auto const& contract = dynamic_cast<ContractDefinition const&>(
+                                *dynamic_cast<Declaration const&>(*_variable.scope()).scope()
+                        );
+                        if (contract.isLibrary()) {
```
Why should this be restricted to libraries only?

Note that for `external` functions, we have a default data location of `calldata` which is not possible to express in source code, so we cannot warn about it.

chase1745 replied:
In the original issue you mentioned "We should do the same at least for the case where the contract is a library.", so I only warned for libraries, but I have changed it to warn for contracts as well.

chase1745 committed:
Added tests for default data location for parameters. Fixed older tests to use explicit storage parameters to silence warnings.

chase1745 commented:
I believe the only issue left is with events. It is warning for no default data location for event parameters, which isn't correct since they are not allowed.
Is there a way to check if the parameter is in an event? @chriseth

chriseth commented:
Please add a function `isEventParameter` in `VariableDeclaration` similar to `isReturnParameter()`. It should be enough to `dynamic_cast` the `scope()` to `EventDeclaration`.

chriseth commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -350,7 +350,26 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
+					if (varLoc == Location::Default)
+					{
+						// Warn users (error in 0.5.0) for using default data location in (return) parameters.
+						if (_variable.sourceUnit().annotation().experimentalFeatures.count(
```
Please create a `bool const v050` variable that holds this condition at the beginning of the function.

chriseth commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -350,7 +350,26 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
 					);
 				if (varLoc == Location::Default || !contract.isLibrary())
+				{
+					if (varLoc == Location::Default)
```
Does this have to be inside the other condition in line 352?

chase1745 replied:
Yes, because of the or condition, a parameter using memory in a public function would be passed to this if statement, so we need to check if it's using the default location. This is a little confusing, but I didn't want to refactor too much of the existing code.

chase1745 committed:
Added isEventParameter to tell if parameter is being called from an event.

axic commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -304,6 +304,9 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
 		using Location = VariableDeclaration::Location;
 		Location varLoc = _variable.referenceLocation();
 		DataLocation typeLoc = DataLocation::Memory;
+		const bool v050 = bool(
```
No need for an explicit typecast here. Also usually we write this as `bool const v050`.

axic commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -350,7 +353,25 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
+				{
+					if (varLoc == Location::Default && !_variable.isEventParameter())
+					{
+						// Warn users (error in 0.5.0) for using default data location in (return) parameters.
```
No need for this comment.

axic commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -350,7 +353,25 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
+						else
+							m_errorReporter.warning(
+									_variable.location(),
+									"Parameter is declared as memory. "
```
Please enclose `memory` in quotes. Perhaps reword it as `Parameter location is assumed as "memory".`

axic commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -368,13 +389,28 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
+						else
+							m_errorReporter.warning(
+									_variable.location(),
+									"Parameter is declared as memory. "
```
The same comments apply to this block as to the above block.

axic commented on libsolidity/ast/AST.cpp:
```diff
@@ -480,6 +480,11 @@ bool VariableDeclaration::canHaveAutoType() const
 
+bool VariableDeclaration::isEventParameter() const
+{
+	return (bool)dynamic_cast<EventDefinition const*>(scope());
```
Please use `!= nullptr` instead of the typecasting.

axic commented:
Please add a change log entry, rebase and squash some of the commits (which belong together, i.e. the test changes).

chase1745 commented:
Just pushed a squashed version of commits. 
Where should this go in the change log since 0.4.24 was just released? 0.5.0?

chriseth commented:
@chase1745 yes, please put this inside the 0.5.0 section.

chase1745 commented:
Added entry to changelog.

axic commented:
@chase1745 sorry for not responding earlier. This is failing test, mostly in `syntaxTests`, which means you need to `isoltest` to regenerate the expectations (since you've added the warning).

axic commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -350,7 +351,24 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
+									"for parameters in publicly visible functions."
+							);
+						else
+							m_errorReporter.warning(
```
Since this is listed as a breaking change, there is no need to have the non-0.5.0 mode here.

axic commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -375,13 +393,28 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
+									"\"memory\" or \"storage\" for parameters."
+							);
+						else
+							m_errorReporter.warning(
```
Same here.

chase1745 commented:
I removed checking for non 0.5.0 mode since this is now a breaking change. I also fixed the new tests to represent this and fixed other non-related tests to not use default data location so they could run as intended without receiving this new error. All isoltests are passing for me now.

chase1745 commented:
Quite a few non-syntax tests are failing now since its a pretty big change. I'll be going through and fixing what I can find over the weekend.

chase1745 committed:
Fixed older tests to use explicit "memory" data location to silence errors.

chase1745 commented:
Still working on them, there's a good number of failures because of this. There also seems to be quite a few external tests throwing errors because of this, not sure what to do about those?

axic commented:
@chase1745 regarding the external tests:
- please fix solc-js and create a PR
- try to fix zeppelin and create a PR against https://github.com/axic/openzeppelin-solidity (against the `solidity-050` branch)
- ignore gnosis failures for the moment

chase1745 committed:
Added documentation for removing default data locations for function parameters.

chase1745 commented:
Submitted PR's for solc-js (https://github.com/ethereum/solc-js/pull/228) and zeppelin against your fork (https://github.com/axic/openzeppelin-solidity/pull/5), and fixed all internal tests, so once those two get merged, I believe all tests should be passing. 

I also updated the documentation to reflect the changes, but I'm not sure how it sounds, so let me know if I should make changes, or if you would like to edit the documentation feel free.

chase1745 commented:
I believe I've fixed all internal tests, but I'm getting 1 weird error on the linux tests. It's happening in `SolidityEndToEndTest.cpp` starting on line `11155`. I can't seem to find what the issue is, everything looks okay to me, any ideas @axic or @chriseth?

I submitted PR's for solc-js (https://github.com/ethereum/solc-js/pull/228) and zeppelin against your fork (https://github.com/axic/openzeppelin-solidity/pull/5).

I also updated the documentation to reflect the changes, but I'm not sure how it sounds, so let me know if I should make changes, or if you would like to edit the documentation feel free.

axic commented:
Thanks! Will a have look shortly.

erak commented:
@chase1745 I've retriggered the tests, but the OpenZeppelin contracts are still failing to compile in ``RBAC.sol`` with ``TypeError: Location must be specified as either "memory" or "storage" for parameters.`` Specifying the data location in ``modifier onlyRole(string roleName)`` should fix this.

chase1745 commented:
@bit-shift thanks, yeah must've just missed that one in my OpenZeppelin PR, I'll fix that.

chase1745 commented:
I rebased from `develop` and also moved tests to more appropriate locations. All tests should be passing once https://github.com/axic/openzeppelin-solidity/pull/8 is merged.

chase1745 commented:
I rebased from `develop` and also moved tests to more appropriate locations. All tests should be passing once https://github.com/axic/openzeppelin-solidity/pull/8 is merged.

christianparpart commented on libsolidity/analysis/ReferencesResolver.cpp:
```diff
@@ -301,6 +301,7 @@ void ReferencesResolver::endVisit(VariableDeclaration const& _variable)
 		using Location = VariableDeclaration::Location;
 		Location varLoc = _variable.referenceLocation();
 		DataLocation typeLoc = DataLocation::Memory;
+		bool const v050 = _variable.sourceUnit().annotation().experimentalFeatures.count(ExperimentalFeature::V050);
```
The next release will be version 0.5.0, so there's no need [anymore] to optionally check for this, is it?

erak replied:
Yes, since this is a breaking change, the warning should be turned into an error (and therefor this check can be removed.)

erak reviewed (changes requested):
Please remove the check for ``pragma experimental "v0.5.0"`` and turn the warning into an error in ``ReferencesResolver::endVisit()``

chase1745 commented:
@bit-shift That is the the error for local variables, not related to this PR, so I didn’t change it, thought it may be better in a separate PR. Happy to change it though if you all think it’s fine for this PR.

erak commented:
@chase1745 Ah yeah, right :) Let's do this in another PR. This one is already pretty large.

chase1745 commented:
@bit-shift  Sounds good, I’ll try to submit one this afternoon. Can you re-trigger the tests now that @axic merged in the changes to OpenZeppelin?

erak commented:
@chase1745 The external test is still failing (``SignatureBouncer.sol``). Here's you can run them locally:
1) Download ``soljson.js``, built on Circle in the ``build_emscripten`` step: https://21670-40892817-gh.circle-artifacts.com/0/soljson.js
2) Place the file under ``[your build directory]/libsolc/soljson.js ``
3) Run ``test/externalTests.sh [your build directory]/libsolc/soljson.js `` in the root of Solidity 

Doing this will help to find breaking contracts :)

chase1745 commented:
🙄I'll find them all this time! Thanks for the tip!

chase1745 commented:
@bit-shift I found 2 parameters in that file to fix, and am almost positive that is all of them in OpenZeppelin, but running the tests the way you said, the current `solidity-050` branch passes all tests before adding in these changes. Any ideas why? I want to make sure I get all of them this time.

chriseth commented:
Sorry, this needs to be rebased now. Perhaps it would make sense to split this PR into multiple PRs:
 - one that updates the syntax tests
 - one that updates the end to end tests
 - one that updates the external tests
 - one that enforces the logic

chase1745 commented:
@chriseth Yeah that makes sense I think. I'll work on that and submit some PR's.

chase1745 commented:
@chriseth Submitted PR's for syntax tests (merged), external tests (#4512), and end to end tests (#4489). Once all 3 have been merged I'll submit a PR for the actual logic so it passes tests and I don't have to rebase a bunch of times.

vs77bb commented:
@chase1745 Great work here. We're happy to tip for all the extra effort you put in - excited to pay out soon 👍

axic commented:
@chase1745 can this PR be closed? I think #4518 is the last in the row superseding this  one.
